### PR TITLE
Fix Edited Line Through Alignment

### PR DESCRIPTION
**Pull request by pubudu-ranasinghe**
### Details

This PR fixes an issue where the horizontal line for the word 'edited' in web and desktop is not at the centre unlike in mobile apps

### Fixed Issues

$ https://github.com/Expensify/App/issues/17181
PROPOSAL: https://github.com/Expensify/App/issues/17181#issuecomment-1502145331

### Tests

1. Open any chat in chrome or desktop
2. Send few messages like normal text, emoji, and markdown
3. Edit the messages
4. Disable your network or force offline from preferences
5. Delete the edited messages
6. Verify that the line through in _edited_ label is aligned properly for all message types

- [x] Verify that no errors appear in the JS console

### Offline tests

Bug is related to offline mode itself. No extra steps needed.

### QA Steps

1. Open any chat in chrome or desktop
2. Send few messages like normal text, emoji, and markdown
3. Edit the messages
4. Disable your network or force offline from preferences
5. Delete the edited messages
6. Verify that the line through in _edited_ label is aligned properly for all message types

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

| Before | After |
| --- | --- |
| Web_Chrome_Before | Web_Chrome_Fixed |
| Web_Safari_Before | Web_Safari_Fixed |

</details>

<details>
<summary>Mobile Web - Chrome</summary>

| Before | After |
| --- | --- |
| MobileWeb_Chrome_Before | MobileWeb_Chrome_Fixed |

</details>

<details>
<summary>Mobile Web - Safari</summary>

| Before | After |
| --- | --- |
| MobileWeb_Safari_Before | MobileWeb_Safari_Fixed |

</details>

<details>
<summary>Desktop</summary>

| Before | After |
| --- | --- |
| Desktop_MacOS_Before | Desktop_MacOS_Fixed |

</details>

<details>
<summary>iOS</summary>

| Before | After |
| --- | --- |
| Mobile_iOS_Before | Mobile_iOS_Fixed |

</details>

<details>
<summary>Android</summary>

| Before | After |
| --- | --- |
| Mobile_Android_Before | Mobile_Android_Fixed |

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionItemFragment.js [hunk 1]
```diff
@@ -107,11 +108,12 @@ const ReportActionItemFragment = (props) => {
             // we render it as text, not as html.
             // This is done to render emojis with line breaks between them as text.
             const differByLineBreaksOnly = Str.replaceAll(html, '<br />', '\n') === text;
+            const isWeb = [CONST.PLATFORM.WEB, CONST.PLATFORM.DESKTOP].includes(getPlatform());
```
Comment (sobitneupane) [inline src/pages/home/report/ReportActionItemFragment.js:111]: @pubudu-ranasinghe Why is this necessary? We discourage the use of platform dependent code unless it is absolutely necessary.

Comment (pubudu-ranasinghe) [reply src/pages/home/report/ReportActionItemFragment.js:111]: This issue doesn't happen on native apps. And providing `inline-block` to RN styles apparently throws an error. So needed to contain it only to desktop and web.

Comment (sobitneupane) [reply src/pages/home/report/ReportActionItemFragment.js:111]: What is the error? If it is not supported by react-native, we better look for other solutions.

**Comment by sobitneupane** [comment]
What is the update here? @pubudu-ranasinghe 

Please do consider the timeline mentioned in the [Contributing guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions) and keep us updated.

**Comment by pubudu-ranasinghe** [comment]
Apologies for the delay. I have been spending a lot of time and exploring several solutions but have not yet arrived at a final one. Unfortunately, this issue is proving difficult to fix in a platform-independent manner.

The issue only occurs in Chrome, affecting the web, mobile Chrome, and desktop app. This is due to a bug in Chrome that affects how it handles line-through on descendant elements with varying font sizes. To demonstrate this, I have created a [fiddle here](https://jsfiddle.net/9fymv64x/3), which shows the difference when opened in Chrome and Safari.

| Chrome | Safari |
| --- | --- |
| Screenshot 2023-05-02 at 18 58 43 | Screenshot 2023-05-02 at 18 59 24 |

When a text decoration is applied to an element, it is propagated to descendant elements. In Chrome, if the descendant font sizes are different, as is the case with the message and edited label in Expensify, the line is not rendered properly. To fix this, we need to prevent the text decoration from propagating to descendants and apply its own line-through. We can do this by making the child element for the edited label an inline-block element as [suggested here](https://stackoverflow.com/a/19529299).

However, React Native does not recognize inline-block as a valid property, throwing an error. Therefore, we can only apply this fix behind platform dependant condition. I have explored other options such as wrapping the edited label and message themselves with Views and playing around with flex display. However, changing the element structure affects how messages are rendered in different platforms, which is not ideal and prone to many regression bugs.

Please note that the current state of the PR only shows experimental changes and not the original solution with inline-block. [This commit](https://github.com/Expensify/App/pull/17781/commits/32627e7331fa50256c3f28f49e8b02285281a9b0) displays the original platform dependant fix.

In my opinion, setting inline-block with platform-specific code is the best solution. However, if there are better ideas in the wider community, I wouldn't mind reopening the issue and handing it over to a different proposal.

cc @sobitneupane

**Comment by sobitneupane** [comment]
No worries @pubudu-ranasinghe. 

>In my opinion, setting inline-block with platform-specific code is the best solution.

@pubudu-ranasinghe Let's go with the solution you suggested but for those styles, instead of `getPlatform()` can we use something like https://github.com/Expensify/App/tree/main/src/styles/getContextMenuItemStyles where index.js and index.native.js are used to handle platform dependent code.

**Comment by pubudu-ranasinghe** [comment]
@sobitneupane there was some change in main branch which introduced `editedLabelStyles`. This is similar to what you suggested with `index.js` and `index.native.js`. I've used the same style with some additional fixes, and the code is now much simpler than before. Please take a look at the PR and let me know if you have any feedback. I've updated the PR description with new screenshots that show a side-by-side comparison.

Additionally, this PR addresses the #18042 defect which I reported earlier but has increased in scope due to a recent change in main.

**Comment by sobitneupane** [comment]
@pubudu-ranasinghe This branch has Merge Conflict. Can you please resolve it and let me know.

**Comment by pubudu-ranasinghe** [comment]
@sobitneupane some of the changes that this PR relied on has been reverted back. I believe this code will be added back as the deploy blocker was not related to those changes as mentioned by @Ollyws. Once the main branch is updated we should be good to merge this. More discussion at #15194

**Comment by sobitneupane** [comment]
Sounds good.

**Comment by pubudu-ranasinghe** [comment]
Just to keep this issue updated, the changes needed for this PR will be added back by #18523. After that we should be good to merge this.

**Comment by sobitneupane** [comment]
@pubudu-ranasinghe We should be good to proceed with this PR now.

**Comment by pubudu-ranasinghe** [comment]
@sobitneupane great! Will update this PR soon

**Comment by pubudu-ranasinghe** [comment]
@sobitneupane Updated

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>

Screenshot 2023-05-24 at 12 53 42

</details>

<details>
<summary>Mobile Web - Chrome</summary>

Screenshot 2023-05-24 at 13 20 21

</details>

<details>
<summary>Mobile Web - Safari</summary>
Screenshot 2023-05-24 at 12 58 21


</details>

<details>
<summary>Desktop</summary>
Screenshot 2023-05-24 at 13 13 04

</details>

<details>
<summary>iOS</summary>

Screenshot 2023-05-24 at 13 04 20

</details>

<details>
<summary>Android</summary>
Screenshot 2023-05-24 at 13 17 41

</details>

**Review by sobitneupane — DISMISSED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

File: src/pages/home/report/ReportActionItemFragment.js [hunk 2]
```diff
@@ -109,7 +109,7 @@ const ReportActionItemFragment = (props) => {
             // Only render HTML if we have html in the fragment
             if (!differByLineBreaksOnly) {
                 const isPendingDelete = props.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE && props.network.isOffline;
-                const editedTag = props.fragment.isEdited ? '<edited></edited>' : '';
+                const editedTag = props.fragment.isEdited ? `<edited ${isPendingDelete ? 'deleted="true"' : ''}></edited>` : '';
```
Comment (NikkiWines) [inline src/pages/home/report/ReportActionItemFragment.js:112]: NAB: Do we need `="true"` here or can we just include `deleted` as boolean attribute?

**Review by NikkiWines — CHANGES_REQUESTED** [review]
Couple small comments

File: src/pages/home/report/ReportActionItemFragment.js (hunk 2 above)
Comment (pubudu-ranasinghe) [reply src/pages/home/report/ReportActionItemFragment.js]: Updated

File: src/components/HTMLEngineProvider/HTMLRenderers/EditedRenderer.js [hunk 3]
```diff
@@ -15,13 +15,15 @@ const propTypes = {
 
 const EditedRenderer = (props) => {
     const defaultRendererProps = _.omit(props, ['TDefaultRenderer', 'style', 'tnode']);
+    const isPendingDelete = !!props.tnode.attributes.deleted;
```
[the remark this answers is not in the document]
Comment (pubudu-ranasinghe) [reply src/components/HTMLEngineProvider/HTMLRenderers/EditedRenderer.js]: Updated

File: src/components/HTMLEngineProvider/HTMLRenderers/EditedRenderer.js [hunk 4]
```diff
@@ -15,13 +15,15 @@ const propTypes = {
...
+
```
[the remark this answers is not in the document]
Comment (pubudu-ranasinghe) [reply src/components/HTMLEngineProvider/HTMLRenderers/EditedRenderer.js]: Updated

**Review by NikkiWines — APPROVED** [review]
:shipit:

**Review by sobitneupane — APPROVED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>

Screenshot 2023-05-26 at 10 54 57
</details>
<details>
<summary>Mobile Web - Safari</summary>

Screenshot 2023-05-26 at 10 58 36
</details>
<details>
<summary>iOS</summary>

Screenshot 2023-05-26 at 11 00 16

</details>
